### DISCUSSION TITLE
Chip-only implementation of SX127X

### DISCUSSION THREAD

**Discussion by Ilikemining1** [jgromes/RadioLib#661] [Q&A] [answered]
Hi,

I'm designing a custom PCB and would like to integrate an SX1276 directly.  I've been looking over the reference designs, and it looks like it should be possible to implement it directly on my PCB.  Will that work as intended with RadioLib?

Additionally, I would like to use both APRS transmission and bidirectional data on LoRa at the same time.  The reference designs show both a low frequency and a high frequency output.  Is it possible to use only one SX IC for this, or would two of them be needed?

Thanks!

IlikeTech

**Comment by jgromes** [comment, marked as answer]
If you follow the reference designs in the datasheet, and take care about impedance matching on the RF part, it should be fine - though I would recommend reading up on sub-GHz PCB design practices first.

Regarding the second part of your question - what does "at the same time" mean to you? If it is full-duplex LoRa combined with simultaneous APRS Tx (so ... a triplex I guess?) then the answer is no, SX127x are half-duplex only and by design cannot receive while transmitting and vice-versa. You cannot even implement full-duplex LoRa, only half-duplex. You will need to switch between Rx and Tx.

Also - this is not really an issue, but a general question - converting to discussion.

**Reply by Ilikemining1** [reply]
Okay, perfect regarding the reference designs.  By at the same time, I just mean can I switch between transmitting APRS, transmitting LoRa, or receiving LoRa with one IC.  I don't mean actual full-duplex transmission, sorry that I didn't make that clear.  I'm designing a board for high altitude ballooning, and want to both be able to communicate with the board bidirectionally over 900MHz LoRa, and send the current location via APRS on an interval.  If the easiest way to do that is with two ICs, one set up for 900MHz and the other for VHF, that's not a problem, but I was wondering if it was possible to combine both functions into one IC for cost and part count reduction.

Thanks, sorry for posting in the wrong place.

**Reply by jgromes** [reply]
In that case, you can switch between different modems during runtime, just be aware that:

a) while transmitting APRS you will not be able to receive data and
b) you will need some additional RF hardware (i.e. a splitter) for the different frequency bands, since you will need different antennas for the APRS and LoRa, as they use different frequencies

EDIT: Alternatively, you could use  a dual-band antenna

**Reply by Ilikemining1** [reply]
So I went over the reference designs and came up with something like this, which I think will be easiest to implement in software.  The top chip is the 915MHz LoRa Transciever, and the bottom one should be set up for VHF-transmit-only operation for APRS.  I think that for the clock, the easiest thing to do is to use one 32MHz TXCO.  Does this seem somewhat reasonable?  The lack of ability to receive LoRa at the same time as APRS is transmitting is somewhat of a deal-killer for a single chip design for my application.

Thanks!

<img width="781" alt="Screen Shot 2023-01-14 at 3 30 51 PM" src="https://user-images.githubusercontent.com/17186671/212495589-d34db44c-5a23-4804-b016-549c41b25517.png">

**Reply by jgromes** [reply]
> the easiest thing to do is to use one 32MHz TXCO

Do you intend to run both SX127x from the same crystal? That seems a bit strange, I'm not sure one crystal is going to be sufficient to drive two of them (since the SX1276s are going to be in parallel to the crystal, the capacity of the pins is going to add up).

> The lack of ability to receive LoRa at the same time as APRS is transmitting is somewhat of a deal-killer for a single chip design for my application.

I don't know what your application is, but I would suggest revisiting this requirement. It seems somewhat arbitrary, what's the reason for it?

**Reply by Ilikemining1** [reply]
Yeah, that definitely makes sense.  I can certainly use two TXCOs if necessary.  The latter isn't necessarily hard requirement, more just that it would be annoying that the board won't be listening for control commands over LoRa when the APRS packets are being transmitted.  Since this board (A payload with some sensors and GPS for controlling a high-altitude balloon) is pretty much just a personal project, the added cost of an extra IC really isn't a big deal.  I would assume to use the different radios,  I just create two different instances of the SX1276 module, correct?

Thanks!

**Reply by jgromes** [reply]
> it would be annoying that the board won't be listening for control commands over LoRa when the APRS packets are being transmitted

You can always implement a "listen-before-talk" feature - on the HAB, run channel activity detection (CAD) before sending the APRS packet. This will allow you to detect the pramble of your LoRa control packet, and switch to reception. If the channel is free, you can send the APRS packet. Alternatively, you could implement a feature that will stop the APRS transmissions for a fixed time since the last command packet was received (to allow for easy commanding, but still restarting the APRS automatically). Though I admit having two SX127x can be slightly more robust when done properly (almost beginning to look like a redundant system).

> To use the different radios, I just create two different instances of the SX1276 module

Correct, though make sure your microcontroller has enough flash/RAM to handle both. Or ... just use two MCUs, for full redundancy ;)

**Reply by Ilikemining1** [reply]
Cool, that makes sense.  My design is using an RP2040 with a 16MB flash chip, so I don't think resources will be an issue.  Another thing I just thought of is that I'd like 100mW TX from both if possible, which means that I'd need to use the PA_BOOST pin out on both units.  Also, other than the reference designs, I haven't found something that explains the antenna matching circuitry, unlike some of the Si4XXX chips.  Am I just missing it?

As a side note, I was actually planning on using an Si4063 transmitter for the APRS originally, and turns out you're doing the AFSK modulation in much the same way that I came up with.  Cool to know that that's the correct way to do AFSK with an FSK transmitter, and nice to not have to implement it myself!

Thanks!

**Reply by jgromes** [reply]
> Also, other than the reference designs, I haven't found something that explains the antenna matching circuitry, unlike some of the Si4XXX chips

You could search through Semtech application notes, but other than recommendation I'm afraid that I can't help you much. RF design really isn't my area of expertise.

If you're going to be pulling 100 mW out of both, just make sure your power supply is sufficient.

> Cool to know that that's the correct way to do AFSK with an FSK transmitter

It's a hack - as far as I know, there is no correct way. These sub-GHz ICs were really not designed for ham radio modes.

**Reply by Ilikemining1** [reply]
Yeah, by correct, I mean "Does it's job even though this isn't really supposed to work"!  As far as the power supplies go, it should be fine, as I'm using a 3.3V 2A switching regulator from regular old lithium AA batteries.  Also, I probably won't transmit on both at the same time unless it's replying to a control command over LoRa, so I should be fine anyway!  I think I'll probably assume that the parts in the reference designs are fine for what I need to do.  Thankfully, it also looks like I can just leave the RF connections that I'm not using disconnected completely.  For your library implementation, I need to connect the SPI bus to both ICs, plus a CS, reset, and both DIO0 and 1, correct?

Thanks!

**Reply by jgromes** [reply]
> I need to connect the SPI bus to both ICs, plus a CS, reset, and both DIO0 and 1, correct?

That's right, though if you want to send AFSK APRS, you will need to connect DIO1 and DIO2 on the second SX1276.
